ocaml/libs/xapi-stdext/lib/xapi-stdext-threads/delay_stubs.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2024 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ */
+
+#include <caml/memory.h>
+#include <caml/threads.h>
+#include <caml/custom.h>
+#include <caml/unixsupport.h>
+
+#include <stdbool.h>
+#include <errno.h>
+#include <pthread.h>
+
+typedef struct delay {
+	pthread_mutex_t mtx;
+	pthread_cond_t cond;
```

Comment:
Might be better to have bindings here just for cond_timedwait, and not for mutex+cond together.
That'll make it easier to replace this code with calls to the standard library, if/when pthread_cond_timedwait gets added upstream.

Also we'll probably eventually replace mutexes and conditions with those provided by Picos once we move to OCaml 5,
because blocking on a fiber needs to be done cooperatively (there is no timed wait primitive there, but it can be constructed out of a condition wait + timeout on the fiber), so the solution with pthread_cond_timedwait would only be a temporary one for OCaml 4.

Reply:
Although I'm not sure if you'll have access to the pthread_mutex_t* in order to be able to call pthread_cond_timedwait without modifying the runtime.

Reply:
I though about this but it would require poking inside Ocaml internal structures which could be changed causing quite some nasty bugs. To avoid also time drifts due to time changes or daylight saving you would have to use monotonic times but most current proposal to extend `Condition` module adding a timed wait are not using them. So at the end a specific limited change seems more sensible.

About fiber that would apply to a refactory of a larger code which is IMO out of scope here.

Reply:
Using the monotonic clock might be useful feedback for https://github.com/ocaml/ocaml/pull/12867/files#diff-c67df461999d14ac8ea7977deb3d7d1d23aea00989818da3096731f18bc788e6 (the upstream PR to add pthread_cond_timedwait).

Perhaps that could be done in a separate commit as part of that PR (it needs to be done at condition creation time).

It'll be a while until that code gets accepted into the stdlib, but once it is we can switch to it.

Although note that pthread_cond* is in general broken on glibc > 2.27
https://sourceware.org/bugzilla/show_bug.cgi?id=25847#c65.
We are not affected by this on XS8, but we will be on XS9 and will likely need to patch glibc with the v5 fix from that  ticket (there is a shorter workaround that Ubuntu has deployed for a while, but it appears to be incomplete)

Reply:
Actually the OCaml 5 code already uses MONOTONIC by default.

Reply:
It's weird they (in Ocaml 5) set `pthread_cond_t` to monotonic clock but don't use `pthread_cond_timedwait`, that's the reason you want to set the type of clock. But I was more referring to the Ocaml side interface, not implementation or `pthread_cond_t` setting. In the sense that we (Xapi) use `Mtime` which is not in the standard library to get monotonic times so adding a timed wait to standard `Condition` with monotonic support will require more changes.

Still the issue of poking into `Condition` stays, although there are some public definition that seems to indicate the usage of `pthread_mutex_t` and `pthread_cond_t` still the relationship between `Condition` and `pthread_cond_t` is not set on stone.

About the bug on glibc still pending that apply to standard `Mutex` and `Condition` but not at this specific implementation (it requires multiple waiters which is not our case).